docs/posts/lms-for-data/index.qmd

```diff
@@ -0,0 +1,691 @@
+```
+
+1. Define a Marvin function to translate text
+2. Define a scalar Python UDF to apply the Marvin function to a column
+3. Apply the Marvin function within the UDF
+
+And apply that UDF to our penguins table:
+
+```{python}
+t = (
+    t.mutate(translated_poem=translate_text(t.poem))  # <1>
+    .relocate(t.columns[:5], "translated_poem")  # <2>
+    .cache()  # <3>
+)
+t
+```
+
+1. Apply the UDF by mutating the table with the function
+2. Rearrange the columns we care about to the front
+3. Cache the table to avoid re-running the UDF
+
+## Writing analytical code
+
+Finally, language models can be used to write code. This is more useful with
+systems around them to execute code, feed back error messages, make adjustments,
+and so on. There are numerous pitfalls with language models writing code, but
+they're fairly good at SQL.
+
+### Three approaches
+
+We can think of three approaches to writing analytical code with language
+models:
+
+1. Use LMs in an analytic subroutine
+2. Use LMs to write an analytic subroutine
+3. Use LMs to write analytic code
+
+We've already seen numerous examples of the first approach above. Let's explore
+the second and third approaches.
+
+### Writing an analytic subroutine
+
+We'll define a function that writes an Ibis UDF, similar to what we've used
+above:
+
+```{python}
+@marvin.fn  # <1>
+def _generate_python_function(text: str) -> str:  # <1>
+    """Generate a simple, typed, correct Python function from text."""  # <1>
+
+
+def create_udf_from_text(text: str) -> str:  # <2>
+    """Create a UDF from text."""
+    return f"""
+import ibis
+
+@ibis.udf.scalar.python
+{_generate_python_function(text)}
+""".strip()  # <2>
+```
+
+1. Define a Marvin function to generate a Python function from text
+2. Define a Python function, wrapping a Marvin function with an Ibis UDF
+
+Now we'll create a UDF from text to count the number of vowels in a string:
+
+```{python}
+udf = create_udf_from_text(  # <1>
+    "a function named count_vowels that given an input string, returns an int w/ the number of vowels (y_included as a boolean option defaulted to False). NO DOCSTRING, we don't document our code in this household"  # <1>
+)
+print(udf)
+```
+
+1. Create a UDF from text
+
+And execute that so the UDF is available:
+
+```{python}
+exec(udf)  # <1>
+```
+
+1. Execute the UDF code to make it available to call
+
+Now we can augment our table with an analytic subroutine written by a language
+model:
+
+```{python}
+t = t.mutate(  # <1>
+    species_vowel_count=count_vowels(t.species),  # <1>
+    island_vowel_count=count_vowels(t.island),  # <1>
+).relocate("species", "species_vowel_count", "island", "island_vowel_count")  # <2>
+t
+```
+
+1. Apply the UDF by mutating the table with the function
+2. Rearrange the columns we care about to the front
+
+:::{.callout-note}
+You could use a LM to write a subroutine that uses a LM! This is an exercise
+left to the reader.
```

Comment:
I think this requires some kind of turtle emoji or similar 😂 

Or perhaps `It's penguins all the way down`?